services/QuillLMS/app/controllers/snapshots_controller.rb

```diff
@@ -0,0 +1,193 @@
+    {value: "Other", name: "Other"}
+  ]
+
+  before_action :validate_request, only: [:count]
+  before_action :authorize_request, only: [:count]
+
+  def count
+    response = Rails.cache.read(cache_key)
+
+    return render json: response if response
+
+    previous_timeframe_start, current_timeframe_start, timeframe_end = calculate_timeframes
+
+    Snapshots::CacheSnapshotCountWorker.perform_async(cache_key,
+      snapshot_params[:query],
+      current_user.id,
+      {
+        name: snapshot_params[:timeframe],
+        previous_start: previous_timeframe_start,
+        current_start: current_timeframe_start,
+        current_end: timeframe_end
+      },
+      snapshot_params[:school_ids],
+      snapshot_params[:grades])
+
+    render json: { message: 'Generating snapshot' }
+  end
+
+  def options
+    render json: {
+      timeframes: TIMEFRAMES.map { |timeframe| format_timeframe_option(timeframe) },
+      schools: Snapshots::SchoolsOptionsQuery.run(current_user.id),
+      grades: GRADE_OPTIONS
+    }
+  end
+
+  private def validate_request
+    return render json: { error: 'timeframe must be present and valid' }, status: 400 unless timeframe_param_valid?
+
+    return render json: { error: 'school_ids are required' }, status: 400 unless school_ids_param_valid?
+  end
+
+  private def timeframe_param_valid?
+    TIMEFRAMES.map { |t| t[:value] }.include?(snapshot_params[:timeframe])
+  end
+
+  private def school_ids_param_valid?
+    snapshot_params[:school_ids]&.any?
+  end
+
+  private def authorize_request
+    schools_user_admins = current_user.administered_schools.pluck(:id)
+
+    return render json: { error: 'user is not authorized for all specified schools' }, status: 403 unless snapshot_params[:school_ids]&.all? { |param_id| schools_user_admins.include?(param_id.to_i) }
+  end
+
+  private def cache_key
+    previous_timeframe_start, current_timeframe_start, timeframe_end = calculate_timeframes
+
+    Snapshots::CacheKeys.generate_key(snapshots_params[:query],
+      previous_timeframe_start,
+      current_timeframe_start,
+      timeframe_end,
+      snapshot_params.fetch(:school_ids, []),
+      snapshot_params.fetch(:grades, []))
+  end
+
+  private def format_timeframe_option(timeframe)
+    {
+      value: timeframe[:value],
+      name: timeframe[:name],
+      default: (DEFAULT_TIMEFRAME == timeframe[:value])
+    }
+  end
+
+  private def find_timeframe(timeframe_value)
+    TIMEFRAMES.find { |timeframe| timeframe[:value] == timeframe_value }
+  end
+
+  private def calculate_timeframes
+    timeframe = find_timeframe(snapshot_params[:timeframe])
+
+    end_of_yesterday = DateTime.current.end_of_day - 1.day
+
+    if snapshot_params[:timeframe] == 'custom'
+      return [
+        send(timeframe[:previous_start]),
+        send(timeframe[:current_start]),
+        send(timeframe[:current_end])
+      ]
+    end
+
+    [
+      timeframe[:previous_start].call(end_of_yesterday),
```

Comment:
Note, take or leave, but you could consider making the custom timeframe and general proc signature the same so you don't have to branch code here, e.g. 
```ruby
custom_params = snapshot_params.slice(:timeframe_custom_start, :timeframe_custom_end)
timeframe[:previous_start].call(end_of_yesterday, custom_params),

# config
{
      value: 'custom',
      name: 'Custom',
      current_start: proc {|_, params| Date.parse(params[:timeframe_custom_start]) }
      ...
}
```
The procs will be more complicated, so this is a take or leave, but I always like eliminating branches if it's not too complicated.